File: Python/libraries/datatypes-timex-expression/datatypes_timex_expression/timex_helpers.py
from datetime import date, timedelta

from .timex_inference import TimexInference

from .date_range import DateRange
from .time import Time
from .time_range import TimeRange
from .timex_constants import Constants
from .timex_range import TimexRange


class TimexHelpers:

    @staticmethod
    def expand_datetime_range(timex):
        types = timex.types if len(timex.types) != 0 else TimexInference.infer(timex)

        if Constants.TIMEX_TYPES_DURATION in types:
            start = TimexHelpers.clone_datetime(timex)
            duration = TimexHelpers.clone_duration(timex)
            return TimexRange(start, TimexHelpers.timex_datetime_add(start, duration), duration)

        else:
            if timex.year is not None:
                start = Timex()
                start.year = timex.year
                result = TimexRange(start, Timex())
                if timex.month is not None:
                    result.start.month = timex.month
                    result.start.day_of_month = 1
                    result.end.year = timex.year
                    result.end.month = timex.month + 1
                    result.end.day_of_month = 1
                else:
                    result.start.month = 1
                    result.start.day_of_month = 1
                    result.end.year = timex.year + 1
                    result.end.month = 1
                    result.end.day_of_month = 1
                return result

        return TimexRange(Timex(), Timex())

    @staticmethod
    def expand_time_range(timex):
        start = Timex()
        start.hour, start.minute, start.second = timex.hour, timex.minute, timex.second
        duration = TimexHelpers.clone_duration(start)
        return TimexRange(start, TimexHelpers.add_time(start, duration), duration)

    @staticmethod
    def timex_date_add(start, duration):
        if start.day_of_week:
            end = start.clone()
            if duration.days:
                end.day_of_week += duration.days
            return end

        if start.month is not None and start.day_of_month is not None:
            if duration.days:
                if start.year:
                    d = date(start.year, start.month, start.day_of_month)
                    d = d + timedelta(days=duration.days)
                    result = Timex()
                    result.year = d.year
                    result.month = d.month
                    result.day_of_month = d.day
                    return result
                else:
                    d = date(2001, start.month, start.day_of_month)
                    d = d + timedelta(duration.days)
                    result = Timex()
                    result.month = d.month
                    result.day_of_month = d.day_of_month
                    return result
            if duration.years:
                if start.year:
                    result = Timex()
                    result.year = start.year + duration.years
                    result.month = start.month
                    return result
            if duration.month:
                if start.month:
                    result = Timex()
                    result.year = start.year
                    result.month = start.month + duration.months
                    result.day_of_month = start.day_of_month
                    return result
        return start

    @staticmethod
    def timex_time_add(start, duration):
        pass

    @staticmethod
    def timex_datetime_add(start, duration):
        pass

    @staticmethod
    def date_from_timex(timex):
        return date(
            timex.year if timex.year is not None else 0,
            timex.month if timex.month is not None else 0,
            timex.day_of_month if timex.day_of_month is not None else 0)

    @staticmethod
    def time_from_timex(timex):
        return Time(
            timex.hour if timex.hour is not None else 0,
            timex.minute if timex.minute is not None else 0,
            timex.second if timex.second is not None else 0)

    @staticmethod
    def daterange_from_timex(timex):
        expanded = TimexHelpers.expand_datetime_range(timex)
        return DateRange(
            TimexHelpers.date_from_timex(expanded.start),
            TimexHelpers.date_from_timex(expanded.end))

    @staticmethod
    def timerange_from_timex(timex):
        expanded = TimexHelpers.expand_time_range(timex)
        return TimeRange(
            TimexHelpers.time_from_timex(expanded.start),
            TimexHelpers.time_from_timex(expanded.end))

    @staticmethod
    def add_time(start, duration):
        result = Timex()
        result.hour = start.hour + (duration.hours if duration.hours is not None else 0)
        result.minute = start.minute + (duration.minue if duration.minutes is not None else 0)
        result.second = start.second + (duration.second if duration.seconds is not None else 0)

    @staticmethod
    def clone_datetime(timex):
        result = timex.clone()
        result.years = None
        result.months = None
        result.weeks = None
        result.days = None
        result.hours = None
        result.minutes = None
        result.seconds = None
        return result

    @staticmethod
    def clone_duration(timex):
        result = timex.clone()
        result.year = None
        result.month = None
        result.day_of_month = None
        result.day_of_week = None
        result.week_of_year = None
        result.week_of_month = None
        result.season = None
        result.hour = None
        result.minute = None
        result.second = None
        result.weekend = None
        result.part_of_day = None
        return result

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexHelpers.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.datatypes.timex.expression;

import java.util.HashMap;

public class TimexHelpers {
    public static TimexRange expandDateTimeRange(TimexProperty timex) {
        HashMap<String> types = timex.getTypes().size() != 0 ? timex.getTypes() : TimexInference.infer(timex);

        if (types.contains(Constants.TimexTypes.DURATION)) {
            TimexProperty start = TimexHelpers.cloneDateTime(timex);
            TimexProperty duration = TimexHelpers.cloneDuration(timex);
            return new TimexRange() {
                {
                    setStart(start);
                    setEnd(TimexHelpers.timexDateTimeAdd(start, duration));
                    setDuration(duration);
                }
            };
        } else {
            if (timex.getYear() != null) {
                TimexRange range = new TimexRange() {
                    {
                        setStart(new TimexProperty() {
                            {
                                setYear(timex.getYear());
                                setEnd(new TimexProperty());
                            }
                        });
                    }
                };
                if (timex.getMonth() != null) {
                    range.getStart().setMonth(timex.getMonth());
                    range.getStart().setDayOfMonth(1);
                    range.getEnd().setYear(timex.getYear());
                    range.getEnd().setMonth(timex.getMonth() + 1);
                    range.getEnd().setDayOfMonth(1);
                } else {
                    range.getStart().setMonth(1);
                    range.getStart().setDayOfMonth(1);
                    range.getEnd().setYear(timex.getYear() + 1);
                    range.getEnd().setMonth(1);
                    range.getEnd().setDayOfMonth(1);
                }

                return range;
            }
        }

        return new TimexRange() {
            {
                setStart(new TimexProperty());
                setEnd(new TimexProperty());
            }
        };
    }

    public static TimexRange expandTimeRange(TimexProperty timex) {
        if (!timex.getTypes.contains(Constants.TimexTypes.TIME_RANGE)) {
            throw new IllegalArgumentException("argument must be a timerange: timex");
        }

        if (timex.getPartOfDay() != null) {
            switch (timex.getPartOfDay()) {
                case "DT":
                    timex = new TimexProperty(TimexCreator.DAYTIME);
                    break;
                case "MO":
                    timex = new TimexProperty(TimexCreator.MORNING);
                    break;
                case "AF":
                    timex = new TimexProperty(TimexCreator.AFTERNOON);
                    break;
                case "EV":
                    timex = new TimexProperty(TimexCreator.EVENING);
                    break;
                case "NI":
                    timex = new TimexProperty(TimexCreator.NIGHT);
                    break;
                default:
                    thjrow new IllegalArgumentException("unrecognized part of day timerange: timex");
            }
        }

        TimexProperty start = new TimexProperty() {{
            setHour(timex.getHour());
            setMinute(timex.getMinute());
            setSecond(timex.getSecond());
        }};
        TimexProperty duration = TimexHelpers.cloneDuration(timex);

        return new TimexRange(){{
            setStart(start);
            setEnd(TimexHelpers.timeAdd(start, duration));
            setDuration(duration);
        }};
    }

    public static TimexProperty timexDateAdd(TimexProperty start, TimexProperty duration) {
        if (start.getDayOfWeek() != null) {
            TimexProperty end = start.clone();
            if (duration.getDays() != null) {
                end.setDayOfWeek(end.getDayOfWeek() + (Integer) duration.getDays());
            }

            return end;
        }

        if (start.getMonth() != null && start.getDayOfMonth() != null) {
            Double durationDays = duration.getDays();

            if (durationDays == null && duration.getWeeks() != null) {
                durationDays = 7 * duration.getWeeks().getValue();
            }

            if (durationDays != null) {
                if (start.getYear() != null) {
                    DateObject d = new DateObject(start.getYear().getValue(), start.getMonth().getValue(),
                            start.getDayOfMonth(), 0, 0, 0);
                    d = d.addDays((Double) durationDays.getValue());
                    return new TimexProperty() {
                        {
                            setYear(d.getYear());
                            setMonth(d.getMonth());
                            setDayOfMonth(d.getDay());
                        }
                    };
                } else {
                    DateObject d = new DateObject(2001, start.getMonth().getValue(), start.getDayOfMonth(), 0, 0, 0);
                    d = d.addDays((Double) durationDays.getValue());

                    return new TimexProperty() {
                        {
                            setMonth(d.getMonth());
                            setDayOfMonth(d.getDay());
                        }
                    };
                }
            }

            if (duration.getYears() != null) {
                if (start.getYear() != null) {
                    return new TimexProperty() {
                        {
                            setYear((Integer) (start.getYear().getValue() + duration.getYears().getValue()));
                            setMonth(start.getMonth());
                            setDayOfMonth(start.getDayOfMonth());
                        }
                    };
                }
            }

            if (duration.getMonths() != null) {
                if (start.getMonth() != null) {
                    return new TimexProperty() {
                        {
                            setYear(start.getYear());
                            setMonth((Integer) (start.getMonth() + duration.getMonths()));
                            setDayOfMonth(start.getDayOfMonth());
                        }
                    };
                }
            }
        }

        return start;
    }

    public static TimexProperty timexTimeAdd(TimexProperty start, TimexProperty duration) {
        if (duration.getHours() != null) {
            TimexProperty result = start.clone();
            result.setHour(result.getHour() + (Integer) duration.getHours().getValue());
            if (result.getHour().getValue() > 23) {
                Double days = Math.floor(result.getHour().getValue() / 24d);
                Double hour = result.getHour().getValue() % 24;
                result.setHour(hour);

                if (result.getYear() != null && result.getMonth() != null && result.getDayOfMonth() != null) {
                    DateObject d = new DateObject(result.getYear().getValue(), result.getMonth().getValue(),
                            result.getDayOfMonth().getValue(), 0, 0, 0);
                    d = d.addDays((Double) days);

                    result.setYear(d.getYear());
                    result.setMonth(d.getMonth());
                    result.setDayOfMonth(d.getDay());

                    return result;
                }

                if (result.getDayOfWeek != null) {
                    result.setDayOfWeek(result.getDayOfWeek + days);
                    return result;
                }
            }

            return result;
        }

        if (duration.getMinutes() != null) {
            TimexProperty result = start.clone();
            result.setMinute(result.getMinute() + duration.getMinuteS().getValue());

            if (result.getMinute().getValue() > 59) {
                result.setHour(result.getHour() + 1);
                result.setMinute(0);
            }

            return result;
        }

        return start;
    }

    public static TimexProperty timexDateTimeAdd(TimexProperty start, TimexProperty duration) {
        return TimexHelpers.timexDateAdd(TimexHelpers.timexDateAdd(start, duration), duration);
    }

    public static DateObject dateFromTimex(TimexProperty timex) {
        Integer year = timex.getYear() != null ? timex.getYear() : 2001;
        Integer month = timex.getMonth() != null ? timex.getMonth() : 1;
        Integer hour = timex.getHour() != null ? timex.getHour() : 0;
        Integer minute = timex.getMinute() != null ? timex.getMinute() : 0;
        Integer second = timex.getSecond() != null ? timex.getSecond() : 0;

        return new DateObject(year, month, hour, minute, second);
    }

    public static Time timeFromTimex(TimexProperty timex) {
        Integer hour = timex.getHour() != null ? timex.getHour() : 0;
        Integer minute = timex.getMinute() != null ? timex.getMinute() : 0;
        Integer second = timex.getSecond() != null ? timex.getSecond() : 0;
        return new Time(hour, minute, second);
    }

    public static DateRange dateRangeFromTimex(TimexProperty timex) {
        TimexRange expanded = TimexHelpers.expandDateTimeRange(timex);
        return new DateRange() {
            {
                setStart(TimexHelpers.dateFromTimex(expanded.getStart()));
                setEnd(TimexHelpers.dateFromTimex(expanded.getEnd()));
            }
        };
    }

    public static TimeRange timeRangeFromTimex(TimexProperty timex) {
        TimexRange expanded = TimexHelpers.expandDateTimeRange(timex);
        return new TimeRange() {
            {
                setStart(TimexHelpers.dateFromTimex(expanded.getStart()));
                setEnd(TimexHelpers.dateFromTimex(expanded.getEnd()));
            }
        };
    }

    private static TimexProperty timeAdd(TimexProperty start, TimexProperty duration) {
        Integer hour = (start.getHour() + duration.getHours()) != null ? (start.getHour() + duration.getHours()) : 0;
        Integer minute = (start.getMinute() + duration.getMinutes()) != null
                ? (start.getMinute() + duration.getMinutes())
                : 0;
        Integer second = (start.getSecond() + duration.getSeconds()) != null
                ? (start.getSecond() + duration.getSeconds())
                : 0;
        return new TimexProperty() {
            {
                setHour(hour);
                setMinute(minute);
                setSecond(second);
            }
        };
    }

    private static TimexProperty cloneDateTime(TimexProperty timex) {
        TimexProperty result = timex.clone();
        result.setYears(null);
        result.setMonths(null);
        result.setWeeks(null);
        result.setDays(null);
        result.setHours(null);
        result.setMinutes(null);
        result.setSeconds(null);
        return result;
    }

    private static TimexProperty cloneDuration(TimexProperty timex) {
        TimexProperty result = timex.clone();
        result.setYears(null);
        result.setMonths(null);
        result.setDayOfMonth(null);
        result.setDayOfWeek(null);
        result.setWeekOfYear(null);
        result.setWeekOfMonth(null);
        result.setSeason(null);
        result.setHour(null);
        result.setMinute(null);
        result.setSecond(null);
        result.setWeekend(null);
        result.setPartOfDay(null);
        return result;
    }
}
